synapse/storage/databases/main/event_federation.py

```diff
@@ -714,77 +722,157 @@ def _get_auth_chain_difference_txn(
                  */
-                INNER JOIN event_edges as g
-                ON g.event_id = e.event_id
+                INNER JOIN event_edges AS edge
+                ON edge.event_id = event.event_id
                 /**
                  * We find the "oldest" events in the room by looking for
                  * events connected to backwards extremeties (oldest events
                  * in the room that we know of so far).
                  */
-                INNER JOIN event_backward_extremities as b
-                ON g.prev_event_id = b.event_id
-                WHERE b.room_id = ? AND g.is_state is ?
-                GROUP BY b.event_id
+                INNER JOIN event_backward_extremities AS backward_extrem
+                ON edge.prev_event_id = backward_extrem.event_id
+                /**
+                 * We use this info to make sure we don't retry to use a backfill point
+                 * if we've already attempted to backfill from it recently.
+                 */
+                LEFT JOIN event_backfill_attempts AS backfill_attempt_info
+                ON backfill_attempt_info.event_id = backward_extrem.event_id
+                WHERE
+                    backward_extrem.room_id = ?
+                    /* We only care about normal events because TODO: why */
+                    AND edge.is_state is ? /* False */
+                    /**
+                     * We only want backwards extremities that are older than or at
+                     * the same position of the given `current_depth` (where older
+                     * means less than the given depth) because we're looking backwards
+                     * from the `current_depth` when backfilling.
+                     *
+                     *                           current_depth (ignore events that come after this, ignore 2-4)
+                     *                           |
+                     * <oldest-in-time> [0]<--[1]▼<--[2]<--[3]<--[4] <newest-in-time>
+                     */
+                    AND event.depth <= ? /* current_depth */
+                    /**
+                     * Exponential back-off (up to the upper bound) so we don't retry the
+                     * same backfill point over and over. ex. 2hr, 4hr, 8hr, 16hr, etc.
+                     *
+                     * We use `2 << (n - 1)` as a power of 2 equivalent for compatibility
+                     * with SQLite. The left shift equivalent only works with powers of 2
+                     * because left shift is a binary operation (base-2). Otherwise, we
+                     * would use `power(2, n)` or the power operator, `2^n`.
+                     */
+                    AND (
+                        backfill_attempt_info IS NULL
+                        OR ? /* current_time */ >= backfill_attempt_info.last_attempt_ts + least((2 << (backfill_attempt_info.num_attempts - 1)) * ? /* step */, ? /* upper bound */)
+                    )
```

Comment:
The query currently results in `sqlite3.OperationalError: no such column: backfill_attempt_info` but only for SQLite 🤔

Not sure why the left joined table alias is acting like this.

Reply:
It's reproducible with just the following snippet ([dbfiddle](https://www.db-fiddle.com/f/eofX81DQdhiPggK22tMDME/0)). SQLite doesn't like the joined table check against `NULL`. Happens with and without the table alias.

`Query Error: Error: SQLITE_ERROR: no such column: failed_backfill_attempt_info`
```
SELECT event_id FROM events AS event
LEFT JOIN event_failed_backfill_attempts AS failed_backfill_attempt_info USING (event_id)
WHERE
    event_id = '$event1'
    AND failed_backfill_attempt_info IS NULL;
```

Setup:
```
CREATE TABLE IF NOT EXISTS events(
    event_id TEXT NOT NULL,
    PRIMARY KEY (event_id)
);

CREATE TABLE IF NOT EXISTS event_failed_backfill_attempts(
    event_id TEXT NOT NULL,
    PRIMARY KEY (event_id)
);

INSERT INTO events VALUES ('$event1'), ('$event2'), ('$event3');
INSERT INTO event_failed_backfill_attempts VALUES ('$event1');
```

Reply:
Looks like it works if we just use one of the columns from the table instead of the table itself:

```sql
SELECT event_id FROM events AS event
LEFT JOIN event_failed_backfill_attempts AS failed_backfill_attempt_info USING (event_id)
WHERE
    event_id = '$event1'
    AND failed_backfill_attempt_info.event_id IS NULL;
```

Solved for now but curious about docs around this ✅